src/data.jl

```diff
@@ -337,6 +378,64 @@ function get_all_gates_dictionary(params::Dict{String, Any}, elementary_gates::A
     
 end
 
+function get_all_CR_gates(params::Dict{String, Any}, elementary_gates::Array{String,1})
+    CR_gates_ids = findall(x -> startswith(x, "CR"), elementary_gates)
+
+    CR_complex = Dict{String, Any}()
+
+    if length(CR_gates_ids) >= 1 
+        for i=1:length(CR_gates_ids)
+
+            gate_type = elementary_gates[CR_gates_ids[i]]
+            # CRP_12
+            if !(gate_type[3] == 'X' || gate_type[3] == 'Y' || gate_type[3] == 'Z')
+                Memento.error(_LOGGER, "Input CR gate type ($(gate_type[3])) is not supported.")
+            end
+            
+            # This implies that discretizations are the same for all gates, checks only for RX/RY/RZ discretization
+            if isempty(params[string(gate_type[1:3],"_discretization")])
+                Memento.error(_LOGGER, "Empty discretization angles for $(gate_type) gate. Input at least one angle")
+            end        
+
+            CR_complex["$(gate_type)"] = Dict{String, Any}()    
+            CR_complex["$(gate_type)"] = QCO.get_discretized_CR_gates(gate_type, CR_complex[gate_type], collect(params[string(gate_type[1:3],"_discretization")]), params["num_qubits"])
+
+        end
+    end
+
+    return CR_complex
+end
+
+function get_discretized_CR_gates(gate_type::String, CR::Dict{String, Any}, discretization::Array{Float64,1}, num_qubits::Int64)
+
+    R_type = gate_type[2:3]
+
+    if length(discretization) >= 1
+
+        for i=1:length(discretization)
+            if R_type == "RX"
+                CR_discrete = [QCO.CRXGate(discretization[i]), QCO.CRXRevGate(discretization[i])]
```

Comment:
@lockwo  Why do we need both CRXGate and CRXRevGate here? If the user input says CRX_12, are you trying to add CRX_21 also as an elementary gate to optimize on?

Reply:
My reasoning was that I wanted to keep the get_full_sized_gate mostly consistent with these new 2 qubit gates while maintaining consistency with 1 qubit rotations. So I enabled the input to be "CRX_12" and had that function determine whether the normal or rev gate is needed. This could be reverse so the caller (be it the code or the user) needs to determine whether the rev or normal gate is necessary and the get_full_sized_gate just returns what it is provided.

Reply:
Ok - I see why you are evaluating both versions here. Instead, do you think this can be directly evaluated inside get_full_sized_gate, so that this redundant evaluation can be avoided if CRXRevGate is not invoked in the user input at all?

Reply:
Partly, I see why it's not a good idea to evaluate this inside since discretization angle also will have to passed in. Can you do this check here itself and evaluate either of the matrices depending on the input and then pass it along to get_full_sized gate?

Reply:
Yes, I'll make the evaluation there, then just return the matrix (or do necessary modifications).

Reply:
Great!